tests/expressions/test_visitor.py

```diff
@@ -0,0 +1,438 @@
+from dataclasses import dataclass
+from decimal import Decimal
+from typing import Any, List
+
+import pytest
+
+from daft.expressions import (
+    Expression,
+    ExpressionVisitor,
+    col,
+    list_,
+    lit,
+)
+from daft.logical.schema import DataType
+from daft.series import Series
+
+
+@dataclass
+class Trace:
+    """Trace tells us which method was called."""
+
+    attr: str
+    args: list[Any]
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Trace):
+            return False
+        if self.attr != other.attr:
+            return False
+
+        # recursive equals using the PyExpr._eq method
+        def eq(obj1: object, obj2: object) -> bool:
+            if isinstance(obj1, Expression) and isinstance(obj2, Expression):
+                return obj1._expr._eq(obj2._expr)
+            elif isinstance(obj1, list) and isinstance(obj2, list):
+                if len(obj1) != len(obj2):
+                    return False
+                for lc, rc in zip(obj1, obj2):
+                    if not eq(lc, rc):
+                        return False
+                return True
+            else:
+                return obj1 == obj2
+
+        return eq(self.args, other.args)
+
+
+class TracingVisitor(ExpressionVisitor[List[Trace]]):
```

Comment:
Shouldn't this inherit from PredicateVisitor instead?

Reply:
Technically yes, practically no difference afaik since it's like a protocol. I've updated because you're right that it _should_

Reply:
Thanks, will approve